packages/service-library/src/servicelib/fastapi/docker_utils.py

```diff
@@ -92,3 +97,48 @@ async def retrieve_image_layer_information(
             except ValidationError:
                 return parse_obj_as(DockerImageManifestsV2, json_response)
     return None
+
+
+_DEFAULT_MIN_IMAGE_SIZE: Final[ByteSize] = parse_obj_as(ByteSize, "50MiB")
+
+
+async def pull_images(
+    images: set[DockerGenericTag],
+    registry_settings: RegistrySettings,
+    progress_cb: AsyncReportCB,
+    log_cb: LogCB,
+) -> None:
+    images_layer_information = await asyncio.gather(
+        *[
+            retrieve_image_layer_information(image, registry_settings)
+            for image in images
+        ]
+    )
+    progress_step_weights = [
+        float(i.layers_total_size) if i else float(_DEFAULT_MIN_IMAGE_SIZE)
+        for i in images_layer_information
+    ]
+    _logger.debug("images to pull sizes: %s", progress_step_weights)
+
+    async with ProgressBarData(
+        num_steps=len(images),
+        step_weights=progress_step_weights,
+        progress_report_cb=progress_cb,
+    ) as pbar:
+        for image, image_layer_info in zip(
+            images, images_layer_information, strict=True
+        ):
+            # NOTE: use gather call here to pull faster
+            # problem with progress in concurrent calls, needs to fix the progress bar
+
+            await pull_image(
```

Comment:
Hmm, very interesting. 
At some point I discussed with @pcrespov if it would be possible to do this in a "fake" way as follows: We would have a simple statistical model which predicts how long it takes to pull images of a given size (perhaps with some added parameters, e.g. if its a hot/cold machine). We would determine the parameters of the model by doing some benchmarking. Then the progress bare would just be a progress bar running through from 0% to 100% in the estimated amount of time. I mention it here because I think it might be worth experimenting with at some point. It would give a very smooth progress bar and remove a lot of logic. Of course we would have to experiment a bit with cooking up the model. Just a remark... 😄